api/src/main/java/org/pmiops/workbench/iam/IamServiceImpl.java

```diff
@@ -0,0 +1,99 @@
+package org.pmiops.workbench.iam;
+
+import static org.pmiops.workbench.iam.SamConfig.SAM_END_USER_GOOGLE_API;
+
+import com.google.api.services.iam.v1.model.Binding;
+import com.google.api.services.iam.v1.model.Policy;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import javax.inject.Provider;
+import org.pmiops.workbench.exceptions.ServerErrorException;
+import org.pmiops.workbench.google.CloudIamClient;
+import org.pmiops.workbench.google.CloudResourceManagerService;
+import org.pmiops.workbench.sam.api.GoogleApi;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+
+@Service
+public class IamServiceImpl implements IamService {
+
+  private final SamApiClientFactory samApiClientFactory;
+  private final CloudIamClient cloudIamClient;
+  private final CloudResourceManagerService cloudResourceManagerService;
+  private final Provider<GoogleApi> endUseGoogleApiProvider;
+  private final SamRetryHandler samRetryHandler;
+
+  @Autowired
+  public IamServiceImpl(
+      SamApiClientFactory samApiClientFactory,
+      CloudIamClient cloudIamClient,
+      CloudResourceManagerService cloudResourceManagerService,
+      @Qualifier(SAM_END_USER_GOOGLE_API) Provider<GoogleApi> endUseGoogleApiProvider,
+      SamRetryHandler samRetryHandler) {
+    this.samApiClientFactory = samApiClientFactory;
+    this.cloudIamClient = cloudIamClient;
+    this.cloudResourceManagerService = cloudResourceManagerService;
+    this.endUseGoogleApiProvider = endUseGoogleApiProvider;
+    this.samRetryHandler = samRetryHandler;
+  }
+
+  @Override
+  public void grantWorkflowRunnerRole(String googleProject) {
+    String petServiceAccountName =
+        getOrCreatePetServiceAccount(googleProject, endUseGoogleApiProvider.get());
+    grantServiceAccountUserRole(googleProject, petServiceAccountName);
+    grantLifeScienceRunnerRole(googleProject, petServiceAccountName);
+  }
+
+  @Override
+  public void grantWorkflowRunnerRoleAsService(String googleProject, String userEmail) {
+    GoogleApi googleApiAsImpersonatedUser = new GoogleApi();
+    try {
+      googleApiAsImpersonatedUser.setApiClient(
```

Comment:
Checking: are you confident there is no other mechanism for retrieving the pet SA for a user? Is the user themselves really the only one who has access to this?

Reply:
I don't know other better ways for this, asked CI in https://trello.com/c/ve2yLn75/115-get-petsa-using-aou-app-service-account
I will ping that thread again.

Reply:
Seems that is the only option we have...

Reply:
Unfortunate - thanks for checking